templates/al2023/runtime/gpu/initialize-nvidia-clock.sh

```diff
@@ -0,0 +1,45 @@
+#!/usr/bin/env bash
+
+set -o errexit
+set -o nounset
+set -o xtrace
+
+if ! gpu-ami-util has-nvidia-devices; then
+  echo >&2 "no NVIDIA devices are present, nothing to do!"
+  exit 0
+fi
+
+# gpu boost clock
+if command -v nvidia-smi &> /dev/null; then
+  echo >&2 "INFO: nvidia-smi found"
+
+  nvidia-smi -q > /tmp/nvidia-smi-check
+  if [[ "$?" == "0" ]]; then
+    sudo nvidia-smi -pm 1 # set persistence mode
+    sudo nvidia-smi --auto-boost-default=0
+
+    GPUNAME=$(nvidia-smi -L | head -n1)
+    echo >&2 "INFO: GPU name: $GPUNAME"
+
+    # set application clock to maximum
+    if [[ $GPUNAME == *"A100"* ]]; then
+      nvidia-smi -ac 1215,1410
+    elif [[ $GPUNAME == *"V100"* ]]; then
+      nvidia-smi -ac 877,1530
+    elif [[ $GPUNAME == *"K80"* ]]; then
+      nvidia-smi -ac 2505,875
+    elif [[ $GPUNAME == *"T4"* ]]; then
+      nvidia-smi -ac 5001,1590
+    elif [[ $GPUNAME == *"M60"* ]]; then
+      nvidia-smi -ac 2505,1177
+    elif [[ $GPUNAME == *"H100"* ]]; then
+      nvidia-smi -ac 2619,1980
```

Comment:
if we aren't going to dynamically pull these at boot, we should proactively add all known cards - `L40S` (g6e) was just launched - `B200`, `GB200`, etc.?

Reply:
also, do we need to set `nvidia-smi -lgc (--lock-gpu-clocks) xx` and `nvidia-smi -lmc (--lock-memory-clocks) yy`? do we test and validate that the clock speeds are maintained across reboots?

Reply:
> if we aren't going to dynamically pull these at boot, we should proactively add all known cards - `L40S` (g6e) was just launched - `B200`, `GB200`, etc.?

I'll go over the existing instances and make sure we have all cards.




> also, do we need to set `nvidia-smi -lgc (--lock-gpu-clocks) xx` and `nvidia-smi -lmc (--lock-memory-clocks) yy`? do we test and validate that the clock speeds are maintained across reboots?

IIUC from `nvidia-smi` help, the flag that we currently set ( `--applications-clocks`) 
specifies the clocks when there is something running on the GPU. While the flags you mention are setting gpu and mem clocks at all times (with or without load). We don't use lock-gpu and lock-mem for the other GPU AMI that we support, so if you think it's beneficial, we can set them.

I'll check the reboot persistence, because this is executed as systemd unit, it should persist across reboots.

Reply:
Clock settings persist reboot, I left out the lgc and lmc flags.

Reply:
sounds good to me!